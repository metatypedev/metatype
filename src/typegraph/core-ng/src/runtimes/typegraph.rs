// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use common::typegraph::Materializer;
use indexmap::IndexMap;

use crate::{
    conversion::runtimes::MaterializerConverter, errors::Result, typegraph::TypegraphContext,
    wit::runtimes::Effect,
};

#[derive(Clone, Debug)]
pub enum TypegraphOperation {
    Resolver,
    GetType,
    GetSchema,
}

impl MaterializerConverter for TypegraphOperation {
    fn convert(
        &self,
        c: &mut TypegraphContext,
        runtime_id: u32,
        effect: Effect,
    ) -> Result<Materializer> {
        let runtime = c.register_runtime(runtime_id)?;

        Ok(Materializer {
            name: match self {
                Self::Resolver => "resolver",
                Self::GetType => "getType",
                Self::GetSchema => "getSchema",
            }
            .to_string(),
            runtime,
            effect: effect.into(),
            data: IndexMap::new(),
        })
    }
}
