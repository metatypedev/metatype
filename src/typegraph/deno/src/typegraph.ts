// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

import * as t from "./types.ts";
import { core } from "./gen/typegraph_core.js";
import { caller, dirname, fromFileUrl } from "./deps/mod.ts";
import { InjectionValue } from "./utils/type_utils.ts";
import {
  serializeFromParentInjection,
  serializeGenericInjection,
  serializeStaticInjection,
} from "./utils/injection_utils.ts";
import { Auth, Cors as CorsWit, Rate, wit_utils } from "./wit.ts";
import { getPolicyChain } from "./types.ts";
import { Artifact, SerializeParams } from "./gen/typegraph_core.d.ts";
import { Manager } from "./tg_manage.ts";
import { log } from "./io.ts";
import { hasCliEnv } from "./envs/cli.ts";
import process from "node:process";

type Exports = Record<string, t.Func>;

type Cors = Partial<CorsWit>;

interface TypegraphArgs {
  name: string;
  dynamic?: boolean;
  builder: TypegraphBuilder;
  prefix?: string;
  secrets?: Array<string>;
  cors?: Cors;
  rate?: Rate;
  disableAutoSerialization?: boolean;
}

export class ApplyFromArg {
  constructor(public name: string | null, public type: number | null) {}
}

export class ApplyFromStatic {
  constructor(public value: any) {}
}

export class ApplyFromSecret {
  constructor(public key: string) {}
}

export class ApplyFromContext {
  constructor(public key: string | null, public type: number | null) {}
}

export class ApplyFromParent {
  constructor(public typeName: string) {}
}

const InjectionSource = {
  asArg: (name?: string, type?: t.Typedef): ApplyFromArg =>
    new ApplyFromArg(name ?? null, type?._id ?? null),
  set: (value: any): ApplyFromStatic => new ApplyFromStatic(value),
  fromSecret: (key: string): ApplyFromSecret => new ApplyFromSecret(key),
  fromContext: (key: string, type?: t.Typedef): ApplyFromContext =>
    new ApplyFromContext(key, type?._id ?? null),
  fromParent: (typeName: string): ApplyFromParent =>
    new ApplyFromParent(typeName),
} as const;

type InjectionSourceType = typeof InjectionSource;

export interface TypegraphBuilderArgs extends InjectionSourceType {
  expose: (
    exports: Exports,
    defaultPolicy?: t.PolicySpec | Array<t.PolicySpec>,
  ) => void;
  inherit: () => InheritDef;
  rest: (graphql: string) => number;
  auth: (value: Auth | RawAuth) => number;
  ref: (name: string) => t.Typedef;
  configureRandomInjection: (params: { seed: number }) => void;
}

export class InheritDef {
  public payload: string | undefined;

  /** inject static value */
  set(value: InjectionValue<unknown>): InheritDef {
    this.payload = serializeStaticInjection(value);
    return this;
  }

  /** inject a value (generic) */
  inject(value: InjectionValue<string>): InheritDef {
    this.payload = serializeGenericInjection("dynamic", value);
    return this;
  }

  /** inject from context */
  fromContext(value: InjectionValue<string>): InheritDef {
    this.payload = serializeGenericInjection("context", value);
    return this;
  }

  /** inject from secret */
  fromSecret(value: InjectionValue<string>): InheritDef {
    this.payload = serializeGenericInjection("secret", value);
    return this;
  }

  /** inject from parent */
  fromParent(value: InjectionValue<string>): InheritDef {
    this.payload = serializeFromParentInjection(value);
    return this;
  }

  /** inject from random */
  fromRandom(): InheritDef {
    this.payload = serializeGenericInjection("random", null);
    return this;
  }
}

export type TypegraphBuilder = (g: TypegraphBuilderArgs) => void;

export class RawAuth {
  constructor(readonly jsonStr: string) {}
}

export interface TypegraphOutput {
  serialize: (config: SerializeParams) => TgFinalizationResult;
  name: string;
}

export interface TgFinalizationResult {
  tgJson: string;
  ref_artifacts: Artifact[];
}

let counter = 0;

export async function typegraph(
  name: string,
  builder: TypegraphBuilder,
): Promise<TypegraphOutput>;
export async function typegraph(args: TypegraphArgs): Promise<TypegraphOutput>;
export async function typegraph(
  args: Omit<TypegraphArgs, "builder">,
  builder: TypegraphBuilder,
): Promise<TypegraphOutput>;
export async function typegraph(
  nameOrArgs: string | TypegraphArgs | Omit<TypegraphArgs, "builder">,
  maybeBuilder?: TypegraphBuilder,
): Promise<TypegraphOutput> {
  ++counter;
  const args = typeof nameOrArgs === "string"
    ? { name: nameOrArgs }
    : nameOrArgs;

  const { name, dynamic, cors, prefix, rate, secrets } = args;
  const builder = "builder" in args
    ? (args.builder as TypegraphBuilder)
    : maybeBuilder!;

  const file = caller();
  if (!file) {
    throw new Error("Could not determine caller file");
  }
  // node/deno compat tick until MET-236 is landed
  const simpleFile = file.replace(/:[0-9]+$/, "").replace(/^file:\/\//, "");
  const path = fromFileUrl(`file://${simpleFile}`);

  const defaultCorsFields = {
    allowCredentials: true,
    allowHeaders: [],
    allowMethods: [],
    allowOrigin: [],
    exposeHeaders: [],
    maxAgeSec: undefined,
  } as CorsWit;

  const defaultRateFields = {
    localExcess: 0,
  } as { localExcess?: number };

  const tgParams = {
    prefix,
    secrets: secrets ?? [],
    cors: cors ? { ...defaultCorsFields, ...cors } : defaultCorsFields,
    rate: rate ? { ...defaultRateFields, ...rate } : undefined,
  };

  core.initTypegraph({ name, dynamic, path, ...tgParams });

  const g: TypegraphBuilderArgs = {
    expose: (exports, defaultPolicy) => {
      core.expose(
        Object.entries(exports).map(([name, fn]) => [name, fn._id]),
        defaultPolicy ? getPolicyChain(defaultPolicy) : [],
      );
    },
    inherit: () => {
      return new InheritDef();
    },
    rest: (graphql: string) => {
      return wit_utils.addGraphqlEndpoint(graphql);
    },
    auth: (value: Auth | RawAuth) => {
      if (value instanceof RawAuth) {
        return wit_utils.addRawAuth(value.jsonStr);
      }
      return wit_utils.addAuth(value);
    },
    ref: (name: string) => {
      return genRef(name);
    },
    configureRandomInjection: (params: { seed: number }) => {
      return core.setSeed(params.seed);
    },
    ...InjectionSource,
  };

  try {
    builder(g);
  } catch (err) {
    if (err.payload && !err.cause) {
      err.cause = err.payload;
    }
    throw err;
  }

  const ret = {
    serialize(config: SerializeParams) {
      try {
        const [tgJson, ref_artifacts] = core.serializeTypegraph(
          config,
        ) as Array<any>; // FIXME: bad typing?
        const result: TgFinalizationResult = {
          tgJson: tgJson,
          ref_artifacts: ref_artifacts,
        };
        return result;
      } catch (err) {
        const stack = (err as any)?.payload?.stack;
        if (stack) {
          // FIXME: jco generated code throws new Error(object) => prints [Object object]
          throw new Error(stack.join("\n"));
        }
        throw err;
      }
    },
    name,
  } as TypegraphOutput;

  if (hasCliEnv()) {
    const manager = new Manager(ret);
    await manager.run();

    // TODO solve hanging process (stdin??)
    setTimeout(() => {
      if (counter === 0) {
        log.debug("exiting");
        process.exit(0);
      }
    }, 10);
  }

  --counter;

  return ret;
}

/** generate a type reference (by name) */
export function genRef(name: string): t.Typedef {
  const value = core.refb(name, []);
  if (typeof value == "object") {
    throw new Error(JSON.stringify(value));
  }
  return new t.Typedef(value, { name });
}
