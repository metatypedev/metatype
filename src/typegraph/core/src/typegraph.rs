// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::conversion::hash::Hasher;
use crate::conversion::runtimes::{convert_materializer, convert_runtime, ConvertedRuntime};
use crate::conversion::types::TypeConversion;
use crate::global_store::SavedState;
use crate::types::{TypeDef, TypeDefExt, TypeId};
use crate::utils::postprocess::{PostProcessor, TypegraphPostProcessor};
use crate::validation::validate_name;
use crate::Lib;
use crate::{
    errors::{self, Result},
    global_store::Store,
};
use common::typegraph::runtimes::TGRuntime;
use common::typegraph::{
    Materializer, ObjectTypeData, Policy, PolicyIndices, PolicyIndicesByEffect, Queries, TypeMeta,
    TypeNode, TypeNodeBase, Typegraph,
};
use indexmap::IndexMap;
use std::cell::RefCell;
use std::collections::hash_map::Entry;
use std::collections::HashMap;
use std::hash::Hasher as _;
use std::path::{Path, PathBuf};
use std::rc::Rc;

use crate::wit::core::{
    Artifact as WitArtifact, Error as TgError, Guest, MaterializerId, PolicyId, PolicySpec,
    RuntimeId, SerializeParams, TypegraphInitParams,
};

#[derive(Default)]
struct IdMapping {
    types_to_hash: HashMap<u32, u64>,
    hash_to_type: HashMap<u64, u32>,
    runtimes: HashMap<u32, u32>,
    materializers: HashMap<u32, u32>,
    policies: HashMap<u32, u32>,
}

#[derive(Default)]
struct RuntimeContexts {
    prisma_typegen_cache: Rc<RefCell<HashMap<String, TypeId>>>,
}

#[derive(Default)]
pub struct TypegraphContext {
    name: String,
    path: Option<Rc<Path>>,
    meta: TypeMeta,
    types: Vec<Option<TypeNode>>,
    runtimes: Vec<TGRuntime>,
    materializers: Vec<Option<Materializer>>,
    policies: Vec<Policy>,
    mapping: IdMapping,
    runtime_contexts: RuntimeContexts,
    saved_store_state: Option<SavedState>,
}

thread_local! {
    static TG: RefCell<Option<TypegraphContext>> = const { RefCell::new(None) };
}

static TYPEGRAPH_VERSION: &str = "0.0.3";

pub fn with_tg<T>(f: impl FnOnce(&TypegraphContext) -> T) -> Result<T> {
    TG.with(|tg| {
        let tg = tg.borrow();
        tg.as_ref()
            .map(f)
            .ok_or_else(errors::expected_typegraph_context)
    })
}

pub fn with_tg_mut<T>(f: impl FnOnce(&mut TypegraphContext) -> T) -> Result<T> {
    TG.with(|tg| {
        let mut tg = tg.borrow_mut();
        tg.as_mut()
            .map(f)
            .ok_or_else(errors::expected_typegraph_context)
    })
}

pub fn init(params: TypegraphInitParams) -> Result<()> {
    #[cfg(test)]
    eprintln!("Initializing typegraph...");

    TG.with(|tg| {
        if let Some(tg) = tg.borrow().as_ref() {
            Err(errors::nested_typegraph_context(&tg.name))
        } else {
            Ok(())
        }
    })?;

    let mut ctx = TypegraphContext {
        name: params.name.clone(),
        path: Some(PathBuf::from(&params.path).into()),
        meta: TypeMeta {
            version: TYPEGRAPH_VERSION.to_string(),
            queries: Queries {
                dynamic: params.dynamic.unwrap_or(true),
                endpoints: vec![],
            },

            cors: params.cors.into(),
            auths: vec![],
            prefix: params.prefix,
            rate: params.rate.map(|v| v.into()),
            secrets: vec![],
            random_seed: Default::default(),
            artifacts: Default::default(),
        },
        types: vec![],
        saved_store_state: Some(Store::save()),
        ..Default::default()
    };

    // register the deno runtime
    let default_runtime_idx = ctx.register_runtime(Store::get_deno_runtime())?;

    ctx.types.push(Some(TypeNode::Object {
        base: TypeNodeBase {
            config: Default::default(),
            description: None,
            enumeration: None,
            injection: None,
            policies: Default::default(),
            runtime: default_runtime_idx,
            title: params.name,
            as_id: false,
        },
        data: ObjectTypeData {
            properties: IndexMap::new(),
            required: vec![],
        },
    }));

    TG.with(move |tg| {
        tg.borrow_mut().replace(ctx);
    });

    Ok(())
}

pub fn finalize_auths(ctx: &mut TypegraphContext) -> Result<Vec<common::typegraph::Auth>> {
    Store::get_auths()
        .iter()
        .map(|auth| match auth.protocol {
            common::typegraph::AuthProtocol::OAuth2 => {
                let profiler_key = "profiler";
                match auth.auth_data.get(profiler_key) {
                    Some(value) => match value {
                        serde_json::Value::Null => Ok(auth.to_owned()),
                        _ => {
                            let func_store_idx = value
                                .as_number()
                                .ok_or_else(|| "profiler has invalid type index".to_string())
                                .and_then(|n| {
                                    n.as_u64().ok_or_else(|| {
                                        "unable to convert profiler index".to_string()
                                    })
                                })? as u32;

                            let type_idx =
                                ctx.register_type(TypeId(func_store_idx).try_into()?, None)?;

                            let mut auth_processed = auth.clone();
                            auth_processed
                                .auth_data
                                .insert_full(profiler_key.to_string(), type_idx.into());

                            Ok(auth_processed)
                        }
                    },
                    None => Ok(auth.to_owned()),
                }
            }
            _ => Ok(auth.to_owned()),
        })
        .collect::<Result<Vec<_>>>()
}

pub fn serialize(params: SerializeParams) -> Result<(String, Vec<WitArtifact>)> {
    #[cfg(test)]
    eprintln!("Serializing typegraph typegraph...");

    let mut ctx = TG.with(|tg| {
        tg.borrow_mut()
            .take()
            .ok_or_else(errors::expected_typegraph_context)
    })?;

    let auths = finalize_auths(&mut ctx)?;

    let mut tg = Typegraph {
        types: ctx
            .types
            .into_iter()
            .enumerate()
            .map(|(id, t)| {
                t.ok_or_else(|| format!("Unexpected: type {id} was not finalized").into())
            })
            .collect::<Result<Vec<_>>>()?,
        runtimes: ctx.runtimes,
        materializers: ctx.materializers.into_iter().map(|m| m.unwrap()).collect(),
        policies: ctx.policies,
        meta: TypeMeta {
            queries: Queries {
                dynamic: ctx.meta.queries.dynamic,
                endpoints: Store::get_graphql_endpoints(),
            },
            // artifacts: {
            //     let arts = ctx.meta.artifacts;
            //     let mut arts = arts.into_iter().collect::<Vec<_>>();
            //     arts.sort_by_cached_key(|pair| pair.0.clone());
            //     arts.into_iter().collect()
            // },
            artifacts: ctx.meta.artifacts,
            random_seed: Store::get_random_seed(),
            auths,
            ..ctx.meta
        },
        path: None,
        deps: Default::default(),
    };

    tg.meta.prefix.clone_from(&params.prefix);

    let pretty = params.pretty;
    TypegraphPostProcessor::new(params).postprocess(&mut tg)?;

    let artifacts = tg
        .meta
        .artifacts
        .values()
        .cloned()
        .map(Into::into)
        .collect::<Vec<_>>();

    Store::restore(ctx.saved_store_state.unwrap());

    let result = if pretty {
        serde_json::to_string_pretty(&tg)
    } else {
        serde_json::to_string(&tg)
    };

    let result = match result.map_err(|e| e.to_string().into()) {
        Ok(res) => res,
        Err(e) => {
            return Err(e);
        }
    };

    Ok((result, artifacts))
}

fn ensure_valid_export(export_key: String, type_id: TypeId) -> Result<()> {
    match type_id.resolve_ref()?.1 {
        TypeDef::Struct(inner) => {
            // namespace
            for (prop_name, prop_type_id) in inner.iter_props() {
                ensure_valid_export(format!("{export_key}::{prop_name}"), prop_type_id)?;
            }
        }
        TypeDef::Func(_) => {}
        _ => return Err(errors::invalid_export_type(&export_key, &type_id.repr()?)),
    }

    Ok(())
}

pub fn expose(
    fields: Vec<(String, TypeId)>,
    default_policy: Option<Vec<PolicySpec>>,
) -> Result<()> {
    let fields = fields
        .into_iter()
        .map(|(key, type_id)| -> Result<_> {
            let concrete_type = type_id.resolve_ref()?.1;
            let policy_chain = &concrete_type.x_base().policies;
            let has_policy = !policy_chain.is_empty();

            // TODO how to set default policy on a namespace? Or will it inherit
            // the policies of the namespace?
            let type_id: TypeId = match (has_policy, default_policy.as_ref()) {
                (false, Some(default_policy)) => {
                    Lib::with_policy(type_id.into(), default_policy.to_vec())?.into()
                }
                _ => type_id,
            };

            Ok((key, type_id))
        })
        .collect::<Result<Vec<_>>>()?;

    with_tg_mut(|ctx| -> Result<_> {
        let mut root = ctx.types.get_mut(0).unwrap().take().unwrap();
        let root_data = match &mut root {
            TypeNode::Object { data, .. } => data,
            _ => return Err("expect root to be an object".into()),
        };
        let res = fields
            .into_iter()
            .map(|(key, type_id)| -> Result<_> {
                if !validate_name(&key) {
                    return Err(errors::invalid_export_name(&key));
                }
                if root_data.properties.contains_key(&key) {
                    return Err(errors::duplicate_export_name(&key));
                }
                ensure_valid_export(key.clone(), type_id)?;

                let type_idx = ctx.register_type(type_id.try_into()?, None)?;
                root_data.properties.insert(key.clone(), type_idx.into());
                root_data.required.push(key);
                Ok(())
            })
            .collect::<Result<Vec<()>>>();

        ctx.types[0] = Some(root);
        res.map(|_| ())
    })?
}

pub fn set_seed(seed: Option<u32>) -> Result<()> {
    Store::set_random_seed(seed);
    Ok(())
}

impl TypegraphContext {
    pub fn hash_type(&mut self, type_def: TypeDef, runtime_id: Option<u32>) -> Result<u64> {
        let type_id = type_def.id().into();
        if let Some(hash) = self.mapping.types_to_hash.get(&type_id) {
            Ok(*hash)
        } else {
            let mut hasher = Hasher::new();
            type_def.hash_type(&mut hasher, self, runtime_id)?;
            let hash = hasher.finish();
            self.mapping.types_to_hash.insert(type_id, hash);
            Ok(hash)
        }
    }

    pub fn register_type(
        &mut self,
        type_def: TypeDef,
        runtime_id: Option<u32>,
    ) -> Result<TypeId, TgError> {
        let hash = self.hash_type(type_def.clone(), runtime_id)?;

        match self.mapping.hash_to_type.entry(hash) {
            Entry::Vacant(e) => {
                // to prevent infinite loop from circular dependencies,
                // we allocate first a slot in the array for the type with None
                // and register it into the mappings so that any dependency
                // would resolve it as already registered with the right idx.

                let idx = self.types.len();
                e.insert(idx as u32);
                self.types.push(None);

                // let tpe = id.as_type()?;

                let type_node = type_def.convert(self, runtime_id)?;

                self.types[idx] = Some(type_node);
                Ok((idx as u32).into())
            }

            Entry::Occupied(e) => Ok((*e.get()).into()),
        }
    }

    pub fn register_materializer(
        &mut self,
        id: u32,
    ) -> Result<(MaterializerId, RuntimeId), TgError> {
        match self.mapping.materializers.entry(id) {
            Entry::Vacant(e) => {
                let idx = self.materializers.len();
                e.insert(idx as u32);
                self.materializers.push(None);
                let converted = convert_materializer(self, Store::get_materializer(id)?)?;
                let runtime_id = converted.runtime;
                self.materializers[idx] = Some(converted);
                Ok((idx as MaterializerId, runtime_id as RuntimeId))
            }
            Entry::Occupied(e) => {
                let mat_idx = *e.get();
                let mat = self.materializers[mat_idx as usize].as_ref().unwrap();
                Ok((mat_idx, mat.runtime))
            }
        }
    }

    pub fn register_policy_chain(&mut self, chain: &[PolicySpec]) -> Result<Vec<PolicyIndices>> {
        chain
            .iter()
            .map(|p| -> Result<_> {
                Ok(match p {
                    PolicySpec::Simple(id) => PolicyIndices::Policy(self.register_policy(*id)?),
                    PolicySpec::PerEffect(policies) => {
                        PolicyIndices::EffectPolicies(PolicyIndicesByEffect {
                            read: policies
                                .read
                                .as_ref()
                                .map(|id| self.register_policy(*id))
                                .transpose()?,
                            create: policies
                                .create
                                .as_ref()
                                .map(|id| self.register_policy(*id))
                                .transpose()?,
                            delete: policies
                                .delete
                                .as_ref()
                                .map(|id| self.register_policy(*id))
                                .transpose()?,
                            update: policies
                                .update
                                .as_ref()
                                .map(|id| self.register_policy(*id))
                                .transpose()?,
                        })
                    }
                })
            })
            .collect()
    }

    pub fn register_policy(&mut self, id: u32) -> Result<PolicyId> {
        if let Some(idx) = self.mapping.policies.get(&id) {
            Ok(*idx)
        } else {
            let converted = Store::get_policy(id)?.convert(self)?;
            let idx = self.policies.len();
            self.policies.push(converted);
            self.mapping.policies.insert(id, idx as u32);
            Ok(idx as PolicyId)
        }
    }

    pub fn register_runtime(&mut self, id: u32) -> Result<RuntimeId, TgError> {
        if let Some(idx) = self.mapping.runtimes.get(&id) {
            Ok(*idx)
        } else {
            let converted = convert_runtime(self, Store::get_runtime(id)?)?;
            let idx = self.runtimes.len();
            self.mapping.runtimes.insert(id, idx as u32);
            match converted {
                ConvertedRuntime::Converted(rt) => self.runtimes.push(rt),
                ConvertedRuntime::Lazy(lazy) => {
                    // we allocate first a slot in the array, as the lazy conversion might register
                    // other runtimes
                    self.runtimes.push(TGRuntime::Unknown(Default::default()));
                    let rt = lazy(id, idx as u32, self)?;
                    self.runtimes[idx] = rt;
                }
            };
            Ok(idx as RuntimeId)
        }
    }

    pub fn find_type_index_by_store_id(&self, id: TypeId) -> Option<u32> {
        let hash = self.mapping.types_to_hash.get(&id.into())?;
        self.mapping.hash_to_type.get(hash).copied()
    }

    pub fn get_correct_id(&self, id: TypeId) -> Result<u32> {
        let id = id.resolve_ref()?.1.id();
        self.find_type_index_by_store_id(id)
            .ok_or(format!("unable to find type for store id {}", u32::from(id)).into())
    }

    pub fn add_secret(&mut self, name: impl Into<String>) {
        // TODO unicity
        self.meta.secrets.push(name.into());
    }

    pub fn get_prisma_typegen_cache(&self) -> Rc<RefCell<HashMap<String, TypeId>>> {
        Rc::clone(&self.runtime_contexts.prisma_typegen_cache)
    }

    pub fn find_materializer_index_by_store_id(&self, id: u32) -> Option<u32> {
        self.mapping.materializers.get(&id).copied()
    }

    pub fn find_policy_index_by_store_id(&self, id: u32) -> Option<u32> {
        self.mapping.policies.get(&id).copied()
    }
}

pub fn current_typegraph_path() -> Result<Rc<Path>> {
    with_tg(|tg| tg.path.clone().unwrap())
}

pub fn current_typegraph_dir() -> Result<PathBuf> {
    let tg_path = current_typegraph_path()?;
    // TODO error handling
    Ok(tg_path.parent().unwrap().to_owned())
}
