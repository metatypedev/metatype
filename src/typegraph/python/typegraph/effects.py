# Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
# SPDX-License-Identifier: MPL-2.0

from enum import Enum, auto

from typegraph.gen.runtimes import (
    Effect,
)


def read() -> Effect:
    return "read"


def create(idempotent: bool = False) -> Effect:
    return {"create": idempotent}


def delete(idempotent: bool = True) -> Effect:
    return {"delete": idempotent}


def update(idempotent: bool = False) -> Effect:
    return {"update": idempotent}


# For injections
class EffectType(Enum):
    CREATE = auto()
    UPDATE = auto()
    DELETE = auto()
    READ = auto()


CREATE = EffectType.CREATE
UPDATE = EffectType.UPDATE
DELETE = EffectType.DELETE
READ = EffectType.READ
