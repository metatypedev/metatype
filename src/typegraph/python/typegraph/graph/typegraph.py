# Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
# SPDX-License-Identifier: MPL-2.0

import inspect
from dataclasses import dataclass
from pathlib import Path
from typing import TYPE_CHECKING, Callable, List, Optional, Union, Any

from typegraph.gen.core import (
    SerializeParams,
    Rate,
    TypegraphInitParams,
)
from typegraph.gen.core import Cors as CoreCors, Rate as CoreRate
from typegraph.gen.utils import Auth
from typegraph.graph.params import Cors, RawAuth
from typegraph.graph.shared_types import FinalizationResult, TypegraphOutput
from typegraph.policy import Policy, PolicyPerEffect, PolicySpec, get_policy_chain
from typegraph.envs.cli import CLI_ENV
from typegraph.sdk import core, sdk_utils
from typegraph.io import Log

if TYPE_CHECKING:
    from typegraph import t


# ExposeItem = Union["t.func", Dict[str, "ExposeItem"]]
ExposeItem = Union["t.func", "t.struct"]


class Typegraph:
    name: str
    dynamic: Optional[bool]
    path: str
    _context: List["Typegraph"] = []
    rate: Optional[CoreRate]
    cors: Optional[CoreCors]
    prefix: Optional[str]

    def __init__(
        self,
        name: str,
        dynamic: Optional[bool] = None,
        *,
        rate: Optional[Rate] = None,
        cors: Optional[Cors] = None,
        prefix: Optional[str] = None,
    ):
        self.name = name
        self.dynamic = dynamic
        self.path = str(Path(inspect.stack()[2].filename).resolve())

        self.rate = Rate(**rate.__dict__) if rate else None

        cors = cors or Cors()
        self.cors = CoreCors(**cors.__dict__)
        self.prefix = prefix

    @classmethod
    def get_active(cls) -> Optional["Typegraph"]:
        if len(cls._context) == 0:
            raise Exception("No active typegraph")
        return cls._context[-1]

    def __call__(self, **kwargs: ExposeItem):
        self.expose(**kwargs)

    def expose(
        self,
        default_policy: Optional[PolicySpec] = None,
        **kwargs: ExposeItem,
    ):
        core.expose(
            [(k, v._id) for k, v in kwargs.items()],
            default_policy=get_policy_chain(default_policy) if default_policy else None,
        )


@dataclass
class ApplyFromArg:
    name: Optional[str]


@dataclass
class ApplyFromStatic:
    value: Any


@dataclass
class ApplyFromSecret:
    key: str


@dataclass
class ApplyFromContext:
    key: str


@dataclass
class ApplyFromParent:
    type_name: str


@dataclass
class Graph:
    typegraph: Typegraph

    def expose(
        self,
        default_policy: Optional[Union[Policy, PolicyPerEffect]] = None,
        **kwargs: ExposeItem,
    ):
        self.typegraph.expose(default_policy, **kwargs)

    def inherit(self):
        from typegraph.injection import InheritDef

        return InheritDef()

    def rest(self, graphql: str) -> int:
        return sdk_utils.add_graphql_endpoint(graphql)

    def auth(self, value: Union[Auth, RawAuth]):
        if isinstance(value, RawAuth):
            return sdk_utils.add_raw_auth(value.json_str)
        else:
            return sdk_utils.add_auth(value)

    def ref(self, name: str) -> "t.typedef":
        return gen_ref(name)

    def configure_random_injection(self, seed: int):
        core.set_seed(seed)

    def as_arg(self, name: Optional[str] = None):
        return ApplyFromArg(name)

    def set(self, value: Any):
        return ApplyFromStatic(value)

    def from_secret(self, key: str):
        return ApplyFromSecret(key)

    def from_context(self, key: str):
        return ApplyFromContext(key)

    def from_parent(self, type_name: str):
        return ApplyFromParent(type_name)


def typegraph(
    name: Optional[str] = None,
    *,
    dynamic: Optional[bool] = None,
    rate: Optional[Rate] = None,
    cors: Optional[Cors] = None,
    prefix: Optional[str] = None,
) -> Callable[[Callable[[Graph], None]], TypegraphOutput]:
    def decorator(builder: Callable[[Graph], None]) -> TypegraphOutput:
        if name is None:
            import re

            # To kebab case
            actual_name = re.sub("_", "-", builder.__name__)
        else:
            actual_name = name

        if CLI_ENV is not None:
            filter = CLI_ENV.filter
            if filter is not None and actual_name not in filter:
                Log.debug("typegraph '{actual_name}' skipped")

                def serialize(params: SerializeParams):
                    raise Exception("typegraph was filtered out")

                return TypegraphOutput(name=actual_name, serialize=serialize)

        tg = Typegraph(
            name=actual_name,
            dynamic=dynamic,
            rate=rate,
            cors=cors,
            prefix=prefix,
        )

        Typegraph._context.append(tg)

        default_cors = CoreCors(
            allow_credentials=True,
            allow_headers=[],
            allow_methods=[],
            allow_origin=[],
            expose_headers=[],
            max_age_sec=None,
        )

        core.init_typegraph(
            TypegraphInitParams(
                name=tg.name,
                dynamic=tg.dynamic,
                path=tg.path,
                rate=tg.rate,
                cors=tg.cors or default_cors,
                prefix=tg.prefix,
            ),
        )

        try:
            builder(Graph(tg))
        except Exception as e:
            import sys

            sys.stderr.write(f"Error in typegraph '{actual_name}':\n")
            sys.stderr.write(str(e))
            sys.exit(1)

        popped = Typegraph._context.pop()
        assert tg == popped

        # config is only known at deploy time
        def serialize_with_artifacts(
            config: SerializeParams,
        ):
            finalization_result = core.serialize_typegraph(config)
            tg_json, ref_artifacts = finalization_result
            return FinalizationResult(tg_json, ref_artifacts)

        tg_output = TypegraphOutput(name=tg.name, serialize=serialize_with_artifacts)

        from typegraph.graph.tg_manage import Manager

        # run from meta/cli
        if CLI_ENV is not None:
            manager = Manager(tg_output)
            manager.run()

        return tg_output

    return decorator


def gen_ref(name: str) -> "t.typedef":
    res = core.refb(name, None)

    from typegraph.t import typedef

    return typedef(res)
