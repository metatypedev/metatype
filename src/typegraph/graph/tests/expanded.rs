// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use std::{path::Path, process::Command, sync::Arc};

use typegraph::{TypeNode as _, TypeNodeExt as _};

#[test]
fn test_expanded_graph() -> color_eyre::Result<()> {
    let project_dir = Path::new(env!("CARGO_MANIFEST_DIR"));
    eprintln!("project_dir={:?}", project_dir);
    let path = project_dir
        .join("../../../tests/metagen/typegraphs/metagen.ts")
        .canonicalize()?;
    eprintln!("path={:?}", path);
    let output = Command::new("cargo")
        .args(
            "run -p meta-cli -- serialize -1 -f"
                .split(' ')
                .collect::<Vec<_>>(),
        )
        .arg(path)
        .output()?;
    eprintln!("status={}", output.status);
    if !output.status.success() {
        eprintln!("stderr={}", String::from_utf8_lossy(&output.stderr));
        color_eyre::eyre::bail!("error running meta-cli");
    }
    let schema = String::from_utf8(output.stdout)?;
    let schema: tg_schema::Typegraph = serde_json::from_str(&schema)?;
    let schema: Arc<_> = schema.into();

    let tg = typegraph::Typegraph::try_from(schema.clone())?;

    println!("namespaces");
    for (ns, obj) in tg.namespace_objects.iter() {
        println!("    /{}: {}", ns.join("/"), obj.name());
    }

    println!("functions");
    for (idx, func) in tg.functions.iter() {
        println!("    {:2}: {}", idx, func.name());
    }

    println!("input_types");
    for (key, ty) in tg.input_types.iter() {
        println!("    {:?}({}): {}", key, ty.tag(), ty.name());
    }

    println!("output_types");
    for (key, ty) in tg.output_types.iter() {
        println!(
            "    {:?}({}): {} parent_idx={}",
            key,
            ty.tag(),
            ty.name(),
            ty.parent().unwrap().idx()
        );
    }

    Ok(())
}
