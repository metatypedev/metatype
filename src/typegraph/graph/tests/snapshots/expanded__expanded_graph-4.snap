---
source: src/typegraph/graph/tests/expanded.rs
expression: "tg.output_types.iter().map(|(idx, ty)|\n            { (idx, (ty.tag(), ty.name())) }).collect::<BTreeMap<_, _>>()"
---
{
    Ty_3/0: (
        "string",
        "string_3fb09",
    ),
    Ty_4/0: (
        "list",
        "root_one_fn_output",
    ),
    Ty_5/0: (
        "object",
        "Student",
    ),
    Ty_6/0: (
        "integer",
        "Student_id_integer",
    ),
    Ty_7/0: (
        "optional",
        "Student_peers_placeholder_optional",
    ),
}
